src/ppx/ppx_eliom_utils.ml

```diff
@@ -461,8 +469,277 @@ end
+    | {ppat_desc= Ppat_construct _; _} ->
+        AC.unit ()
+    | _ ->
+        print_error ~loc:pattern.ppat_loc Fatal_error
+
+  let apply_function_expr ~loc fun_name args_list =
+    Exp.apply
+      (Exp.ident (Location.mkloc (Longident.parse fun_name) loc))
+      args_list
+
+  let rec args_parser expr arg_list =
+    match expr with
+    | [%expr fun ([%p? pattern] : [%t? typ]) -> [%e? expr']] ->
+        args_parser expr' ((pattern, typ, Asttypes.Nolabel) :: arg_list)
+    | [%expr fun () -> [%e? expr']] ->
+        args_parser expr'
+          ((AC.punit (), AC.tconstr "unit" [], Asttypes.Nolabel) :: arg_list)
+    | { pexp_desc=
+          Pexp_fun
+            ( Asttypes.Labelled s
+            , None
+            , {ppat_desc= Ppat_constraint (pattern, typ)}
+            , expr' ) } ->
+        args_parser expr' ((pattern, typ, Asttypes.Labelled s) :: arg_list)
+    | { pexp_desc=
+          Pexp_fun
+            ( Asttypes.Optional s
+            , None
+            , {ppat_desc= Ppat_constraint (pattern, typ)}
+            , expr' ) } ->
+        args_parser expr' ((pattern, typ, Asttypes.Optional s) :: arg_list)
+    | [%expr fun [%p? p] -> [%e? _]]
+    | { pexp_desc=
+          Pexp_fun ((Asttypes.Optional _ | Asttypes.Labelled _), None, p, _) }
+      ->
+        print_error ~loc:p.ppat_loc Missing_argument_type
+    | e -> (
+      match arg_list with
+      | [] ->
+          print_error ~loc:e.pexp_loc No_arguments
+      | arg_list ->
+          List.rev arg_list )
+
+  let connected_wrapper ~loc rpc_name fun_name args_list _pat_of_args _typ_of_args aux_apply =
+    let apply =
+      apply_function_expr ~loc fun_name
+        (List.map
+           (fun (pattern, _, label) -> (label, pat_to_exp pattern))
+           args_list)
+    in
+    [%expr
+      let aux =
+        ~%(Eliom_client.server_function ~name:[%e rpc_name]
+             [%json: [%t _typ_of_args]]
+             (Os_session.connected_wrapper (fun [%p _pat_of_args] ->
+                  [%e apply])))
+      in
+      [%e aux_apply]]
+
+  let connected_rpc ~id ~loc rpc_name fun_name args_list _pat_of_args _typ_of_args aux_apply =
+    let apply =
+      apply_function_expr ~loc fun_name
+        ( (Asttypes.Nolabel, pat_to_exp id)
+        :: List.map
+             (fun (pattern, _, label) -> (label, pat_to_exp pattern))
+             args_list )
+    in
+    [%expr
+      let aux =
+        ~%(Eliom_client.server_function ~name:[%e rpc_name]
+             [%json: [%t _typ_of_args]]
+             (Os_session.connected_rpc (fun [%p id] [%p _pat_of_args] ->
+                  [%e apply])))
+      in
+      [%e aux_apply]]
+
+  let connected_rpc_o ~id ~loc rpc_name fun_name args_list _pat_of_args _typ_of_args aux_apply
+      =
+    let apply =
+      apply_function_expr ~loc fun_name
+        ( (Asttypes.Nolabel, pat_to_exp id)
+        :: List.map
+             (fun (pattern, _, label) -> (label, pat_to_exp pattern))
+             args_list )
+    in
```

Comment:
this code block seem to be the same as connected rpc. There is maybe a possible factorization.